app/main.py

```diff
@@ -1,34 +1,160 @@
+        self._validate_field()
+
+    def __str__(self) -> str:
+        return f"{self.field}"
+
+    def _validate_field(self) -> None:
+        if len(self.field) != 10:
+            raise ValidationInputDataError(
+                "The total number of ships should be 10"
+            )
+
+        ships_count = {1: 0, 2: 0, 3: 0, 4: 0}
+        for ship_start, ship_end in self.field.keys():
+            ship_len = 1 + max(
+                abs(ship_start[0] - ship_end[0]),
+                abs(ship_start[1] - ship_end[1])
+            )
+            if ship_len > 4:
+                raise ValidationInputDataError(
+                    "The maximum length of the ship should not exceed 4 decks"
+                )
+            ships_count[ship_len] += 1
+
+        if ships_count != {1: 4, 2: 3, 3: 2, 4: 1}:
+            raise ValidationInputDataError(
+                """There should be 4 single-deck ships
+                There should be 3 double-deck ships
+                There should be 2 three-deck ships
+                There should be 1 four-deck ship"""
+            )
+
+        ship_deck_all = []
+        for ship in self.field:
+            ship = sorted(ship)
+            ship_deck = []
+            for vertical in range(ship[0][0], ship[1][0] + 1):
+                for horizontal in range(ship[0][1], ship[1][1] + 1):
+                    spase = [vertical, horizontal]
+                    ship_deck.append(spase)
+            ship_deck_all += ship_deck
+
+        for ship in self.field:
+            ship = sorted(ship)
+            y0 = ship[0][0] - 1 if ship[0][0] - 1 >= 0 else 0
+            y1 = min(ship[1][0] + 1, GRID_SIZE - 1)
+            x0 = ship[0][1] - 1 if ship[0][1] - 1 >= 0 else 0
+            x1 = min(ship[1][1] + 1, GRID_SIZE - 1)
+            ship_spase = []
+            for vertical in range(y0, y1 + 1):
+                for horizontal in range(x0, x1 + 1):
+                    spase = [vertical, horizontal]
+                    ship_spase.append(spase)
+
+            ship_deck = []
+            for vertical in range(ship[0][0], ship[1][0] + 1):
+                for horizontal in range(ship[0][1], ship[1][1] + 1):
+                    spase = [vertical, horizontal]
+                    ship_deck.append(spase)
+
+            for deck in ship_deck:
+                ship_spase.remove(deck)
+
+            for spase in ship_spase:
+                if spase in ship_deck_all:
+                    raise ValidationInputDataError(
+                        "Ships shouldn't be located in the neighboring cells "
+                        "(even if cells are neighbors by diagonal)."
+                    )
+
+    def fire(self, location: tuple) -> str:
+        for coords, ship in self.field.items():
+            ship_decks = ship.get_deck(location[0], location[1])
+            if ship_decks is not None:
+                return ship.fire(location[0], location[1])
+        return "Miss!"
+
+    def print_field(self) -> None:
```

Comment:
Triple loops are not allowed, think about how to rewrite the code with less nesting